src/data_struct/revision_map.rs

```diff
@@ -0,0 +1,223 @@
+use std::{
+    borrow::Borrow,
+    collections::{
+        hash_map::{Entry, OccupiedEntry, VacantEntry},
+        HashMap, LinkedList,
+    },
+    fmt::Debug,
+    hash::Hash,
+};
+
+/// a `RevisionMap` stores linked lists indexed by given keys.
+/// The element inside the linked list are stored in reverse insertion order
+/// while the keys are stored in arbitrary order.
+///
+/// Map {
+///     key2: b
+///     key1: a" -> a' > a
+///     key3: c' -> c
+/// }
+#[derive(Default, Debug, PartialEq, Eq)]
+pub struct RevisionMap<K, V>
+where
+    K: Debug + PartialEq + Eq + Hash,
+    V: Debug,
+{
+    pub(crate) map: HashMap<K, LinkedList<V>>,
+}
+
+impl<K, V> RevisionMap<K, V>
+where
+    K: Hash + PartialEq + Eq + Clone + Debug,
+    V: Clone + Debug,
+{
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+        }
+    }
+
+    #[must_use]
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            map: HashMap::with_capacity(capacity),
+        }
+    }
+
+    /// Returns the number of chains stored.
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    /// Returns the total number of elements stored.
+    pub fn count_elements(&self) -> usize {
+        self.map.values().map(LinkedList::len).sum()
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.map.is_empty()
+    }
+
+    pub fn chain_length(&self, key: &K) -> Option<usize> {
+        self.map.get(key).map(LinkedList::len)
+    }
+
+    fn insert_new_chain(entry: VacantEntry<K, LinkedList<V>>, value: V) {
+        let mut new_chain = LinkedList::new();
+        new_chain.push_front(value);
+        entry.insert(new_chain);
+    }
+
+    fn insert_in_chain(mut entry: OccupiedEntry<K, LinkedList<V>>, value: V) {
+        let chain = entry.get_mut();
+        chain.push_front(value);
+    }
+
+    /// Inserts value at the front of the chain for a given key
+    pub fn insert(&mut self, key: K, value: V) {
+        match self.map.entry(key) {
+            Entry::Occupied(entry) => Self::insert_in_chain(entry, value),
+            Entry::Vacant(entry) => Self::insert_new_chain(entry, value),
+        }
+    }
+
+    /// Returns the last revised value for a given key.
+    pub fn get_current_revision<Q>(&self, key: &Q) -> Option<&V>
```

Comment:
I hesitated for this naming because `last` could refer to the end of the list (which are the older values)...
The Rust LinkedList call this action `get_front`

Reply:
Maybe simply calling it `get` since this is the intended way to use this structure

Reply:
Yes, you are right I didn't think about it... would `get_latest(k)` be less confusing?